app/busy.go

```diff
@@ -25,27 +42,47 @@ func (b *Busy) IsBusy() bool {
 	return atomic.LoadInt32(&b.busy) != 0
 }
 
-// Set marks the server as busy for dur duration.
+// Set marks the server as busy for dur duration and notifies cluster nodes.
 func (b *Busy) Set(dur time.Duration) {
 	b.mux.Lock()
 	defer b.mux.Unlock()
 
-	b.clear()
-	atomic.StoreInt32(&b.busy, 1)
+	// minimum 1 second
+	if dur < (time.Second * 1) {
+		dur = time.Second * 1
+	}
 
+	b.setWithoutNotify(dur)
 	b.timer = time.AfterFunc(dur, b.Clear)
```

Comment:
My only concern is around this nodes restartining inside the HA environment and leaving the environment inconsistent.

My proposal here would be to set "IsBusy" based on the expire date instead of the "busy" flag. This way you don' tneed to call this `Clear` function manually.

Another thing is, when a node is restarted we should get the busy state from the current active node (But I think we can leave that as an improvement for later).

Reply:
Thanks @jespino.  I agree just having an expiry versus flag + expiry is a bit simpler, but at the expense of latency. The flag check is very fast and doesn't not require a sync.lock to check.  I don't yet have a sense of how much load a typical node gets, and while I can image that search doesn't get a large amount of concurrent access, I wonder if things like typing and statuses won't benefit from a sync-lock free implementation, especially considering that 99.99+% of the time the busy flag is not set, but we have to check for every typing/status event.   

Either way, I don't think that will solve the consistency issue after a node restarts.  Currently, I have the node that set the flag be the one that clears it.  Therefore. if the originating node restarts, the other nodes will keep their busy flag past the expiry time - I believe that is the issue you are concerned about, and quite correctly.  I need to let each node clear their own flag after the expiry.  I believe that would address the immediate concern and keep the fast isBusy check.

I will also look into how to get the current state from the cluster after a node restart and see if that is simple enough to add now or backlog it.

Reply:
via https://github.com/mattermost/mattermost-server/pull/13267/commits/dda7ffe3f0fb7524934b03ec7a188d2b0150843c each node now clears its own busy flag after expiry.  

@jespino, if you feel passionate about removing the flag in favor of just using an expiry, and the contention of a sync.lock on isBusy is not an issue, I can remove the flag.  I still like the idea of a sync.lock free isBusy.

Reply:
This perfectly addresses my concern about nodes inconsistency on restart (about not clearing the isBusy state), and keep the performance there pretty high. The other one, I think we can backlog it for now.